### ISSUE TITLE
Clarify Sleuth support

### ISSUE THREAD

**Issue by asarkar** [yidongnan/grpc-spring-boot-starter#389]
It's not at all clear from the documentation what is traced in the presence of Spring Cloud Sleuth. Are tracing headers added to all client-server calls? How's that helpful if the server doesn't return those back? Please update the docs to clarify, preferably using an image like [this](https://raw.githubusercontent.com/spring-cloud/spring-cloud-sleuth/2.2.x/docs/src/main/asciidoc/images/trace-id.png)

**Comment by ST-DDT** [comment]
I'm not familiar with sleuth, so I appreciate any help.

**Comment by asarkar** [comment]
@ST-DDT Are you saying that the code claims to support Sleuth, but you don't know what it does?

**Comment by ST-DDT** [comment]
This sounds a bit offensive to me. But maybe i just used the wrong wording.

I know how it works, but I dont know how (and also what) to describe it in a way thats adds more value to it than a link to sleuth's docs itself.

Setup howto:
Add sleuth similar to your http app
Add the grpc-instrumentation as dependency
Well, thats all you have to do.

Which data are transfered for which purpose?
sleuth/brave/zipkin's docs are centered on that topic, so why should I repeat it?!

**Comment by asarkar** [comment]
I don't know why you were offended, but some people are more sensitive than others. It was obviously not my intent to offend you.

Sleuth works by adding metadata to remote calls. For HTTP, it adds headers; I assume that's what it does for gRPC. What I'm trying to understand here is what does it mean to enable Sleuth dependency. For bidirectional streaming, the client will send several requests, and receive several response. Unless someone owns both client and server side code, and adds Sleuth to the build on both ends, I don't understand how Sleuth will be able to inject metadata. I also don't understand what this library does when Sleuth is on the classpath.

Is my question clearer now? I can, of course, look in the code, but I think this particular section of the docs could use some attention.

**Comment by ST-DDT** [comment]
(I had the feeling **we**/this library were being accused of making a false claim.)
If that wasn't intended, then I don't mind. 

Aside from sleuth you also need [brave-instrumentation-grpc](https://mvnrepository.com/artifact/io.zipkin.brave/brave-instrumentation-grpc). That library contains all the relevant logic. This library only creates the few [relevant beans](https://github.com/yidongnan/grpc-spring-boot-starter/blob/1043e47ca980ac25fe7ce9b01cf3c8aa4932cef2/grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/autoconfigure/GrpcClientTraceAutoConfiguration.java#L53).

Does the following link help you to understand how it works?
https://github.com/openzipkin/brave/tree/master/instrumentation/grpc#message-processing-callback-context

**Comment by asarkar** [comment]
So it turns out that Sleuth gRPC tracing is completely broken. I opened https://github.com/spring-cloud/spring-cloud-sleuth/issues/1749, which was closed because “it wasn’t their problem”. I then opened https://github.com/openzipkin/zipkin-support/issues/40, which was closed because “it is not easy to implement”. I then opened https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1313, which was implemented and seems to be working.

In short, the “Sleuth support” you’re providing is pretty much useless. The only library that traces gRPC properly is OpenTelemetry, and it isn’t released yet.

**Comment by ST-DDT** [comment]
AFAICT its only broken for (Kotlin) coroutines. However, the fact that they have such a problem seems somewhat suspicious to me. I assume that there is simply some kind of (context) bridge missing.

I don't have time to have an in depth look into this anytime soon. Would it be possible for you to write a minimalistic (Java based) test for this? Running it in a Kotlin JRE is probably not an issue.  If a strictly java source code based test doesn't work, then a Kotlin one is fine too, as long as I can use my usual debugging tools.
Then maybe I can figure out where their tracing context is lost, which is the first half of fixing the issue.

**Comment by asarkar** [comment]
It could be that Kotlin coroutine is making things worse, but the fundamental issue is the Brave design that context isn't propagated using gRPC context but their own. OpenTelemetry was going down that route, until they listened to my suggestion, and simply used gRPC context instead. They did nothing special for coroutines, but it works like a charm.

You may close this ticket if you feel like; I have little interest in supporting libraries that simply brush off user problems (Sleuth and Brave).

As an aside, I do believe that cross-cutting concerns like tracing and metrics should be separate from the core library, because people may not be using the same third-party libraries that you assumed they would (like Sleuth or Micrometer).

**Comment by ST-DDT** [comment]
@anjeyy You might have run into this issue: https://github.com/spring-projects/spring-boot/issues/19860
(`org.springframework.boot.context.properties.ConfigurationBeanFactoryMetadata` has been removed/replaced in spring-boot 2.4.0)


Maybe you need a newer version of `spring-cloud-starter-sleuth`?
